classes/oai/PKPOAIDAO.inc.php

```diff
@@ -143,22 +143,13 @@ public function getRecord($dataObjectId, $setIds = [])
      */
     public function getRecords($setIds, $from, $until, $set, $offset, $limit, &$total)
     {
-        $result = $this->_getRecordsRecordSet($setIds, $from, $until, $set);
+        $query = $this->_getRecordsRecordSetQuery($setIds, $from, $until, $set);
+        $total = $query->count();
+        $results = $query->offset($offset)->limit($limit)->get();
```

Comment:
Considering the OAI doesn't have a "pageNumber" parameter and that the result set is already sorted by IDs, it's possible to achieve a better performance by replacing the `offset` (the more you skip, the slower it gets) by a `where` clause (that [Laravel's method](https://laravel.com/api/7.x/Illuminate/Database/Query/Builder.html#method_chunkById) we discussed in a Beacon issue might cover it).

Reply:
That's a good idea, but it'll require us to change the way OAI resumption tokens are stored so that we know the first ID to fetch on subsequent calls. For now the changes here and in https://github.com/pkp/ojs/pull/3134 at least move the iteration out of PHP, which will resolve the bottleneck noted in the FIXME. We can come back to this when it shows up as a new bottleneck.

Reply:
I think the offset/limit will be enough, unless there are installations with millions of records out there 🤔
In case we have access to such installation it would be good to run a query just to check the performance of retrieving the last page.

Reply:
How would we could then get the total numbers?